src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+}
+
+// get RZ cell indices corresponding to point of interest
+bool FieldMapB::GetIndices(double R, double Z, int *idxR, int *idxZ, double *deltaR, double *deltaZ)
+{
+  // boundary check
+  if( R > maxs[0] || R < mins[0] || Z > maxs[1] || Z < mins[1] ) {
+    return false;
+  }
+
+  // get indices
+  *deltaR = std::modf( (R - mins[0]) / steps[0], &ir_d );
+  *deltaZ = std::modf( (Z - mins[1]) / steps[1], &iz_d );
+
+  *idxR = static_cast<int>( ir_d );
+  *idxZ = static_cast<int>( iz_d );
+
+  return true;
+}
+
+// get XYZ cell indices corresponding to point of interest
+bool FieldMapB::GetIndices(double X, double Y, double Z, int *idxX, int *idxY, int *idxZ, double *deltaX, double *deltaY, double *deltaZ)
+{
+  // boundary check
+  if( X > maxs[0] || X < mins[0] || Y > maxs[1] || Y < mins[1] || Z > maxs[2] || Z < mins[2] ) {
+    return false;
+  }
+
+  // get indices
+  *deltaX = std::modf( (X - mins[0]) / steps[0], &ix_d );
+  *deltaY = std::modf( (Y - mins[1]) / steps[1], &iy_d );
+  *deltaZ = std::modf( (Z - mins[2]) / steps[2], &iz_d );
+
+  *idxX = static_cast<int>( ix_d );
+  *idxY = static_cast<int>( iy_d );
+  *idxZ = static_cast<int>( iz_d );
+
+  return true;
+}
+
+// load data
+void FieldMapB::LoadMap(const std::string& map_file, double scale)
+{
+  std::string   line;
+  std::ifstream input(map_file);
+  if (!input) {
+    std::cout << "FieldMapB Error: file \"" << map_file << "\" cannot be read." << std::endl;
+  }
+
+  std::vector<double> coord = {};
+  std::vector<double> Bcomp = {};
+
+  while (std::getline(input, line).good()) {
+    std::istringstream iss(line);
+
+    coord.clear();
+    Bcomp.clear();
+
+    if( case_BrBz ) {
+      coord.resize(2);
+      Bcomp.resize(2);
+      iss >> coord[0] >> coord[1] >> Bcomp[0] >> Bcomp[1];
+
+      if( ! GetIndices( coord[0], coord[1], &ir, &iz, &dr, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_RZ[ ir ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale };
+        //ROOT::Math::XYZPoint p(coord[0], 0, coord[1]);
+        //std::cout << p << " -> " << trans*p << std::endl;
+        //std::cout << indices[0] << ", " << indices[1] << ", " << Bcomp[0] << ", " << Bcomp[1] << std::endl;
+      }
+    }
+    else {
+      coord.resize(3);
+      Bcomp.resize(3);
+      iss >> coord[0] >> coord[1] >> coord[2] >> Bcomp[0] >> Bcomp[1] >> Bcomp[2];
+
+      if( ! GetIndices(coord[0], coord[1], coord[2], &ix, &iy, &iz, &dx, &dy, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_XYZ[ ix ][ iy ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale, Bcomp[2] * scale };
+      }
+    }
+  }
+}
+
+// get field components
+void FieldMapB::fieldComponents(const double* pos, double* field)
+{
+  // coordinate conversion
+  auto p = trans_inv * ROOT::Math::XYZPoint(pos[0], pos[1], pos[2]);
+
+  if( case_BrBz ) {
```

Comment:
Use switch case on enum here for scalability to future BrBzBphi description of MARCO I imagine.